wagtail/contrib/settings/tests/shared/test_preview.py

```diff
@@ -0,0 +1,191 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from wagtail.admin.staticfiles import versioned_static
+from wagtail.models import Site
+from wagtail.test.testapp.models import (
+    PreviewableGenericSetting,
+    PreviewableSiteSetting,
+)
+from wagtail.test.utils import WagtailTestUtils
+
+
+class TestGenericSiteSettingPreview(WagtailTestUtils, TestCase):
+    model = PreviewableGenericSetting
+    other_model = PreviewableSiteSetting
+
+    def setUp(self):
+        self.user = self.login()
+        site = Site.objects.get(is_default_site=True)
+
+        PreviewableGenericSetting.objects.create(
+            text="An initial previewable generic setting"
+        )
+        PreviewableSiteSetting.objects.create(
+            text="An initial previewable site setting",
+            site=site,
+        )
+
+        self.setting = self.model.objects.first()
+        self.app_label = self.model._meta.app_label
+        self.model_name = self.model._meta.model_name
+        self.verbose_name = self.model._meta.verbose_name
+        self.preview_on_edit_url = reverse(
+            "wagtailsettings:preview_on_edit",
+            args=(self.app_label, self.model_name, self.setting.pk),
+        )
+        self.session_key_prefix = f"wagtail-preview-{self.app_label}-{self.model_name}"
+        self.edit_session_key = f"{self.session_key_prefix}-{self.setting.pk}"
+
+        self.post_data = {
+            "text": f"An edited {self.verbose_name}",
+        }
+
+    def test_preview_on_edit_with_valid_then_invalid_data(self):
+        response = self.client.post(self.preview_on_edit_url, self.post_data)
+
+        # Check the JSON response
+        self.assertEqual(response.status_code, 200)
+        self.assertJSONEqual(
+            response.content.decode(),
+            {"is_valid": True, "is_available": True},
+        )
+
+        # Send an invalid update request
+        response = self.client.post(self.preview_on_edit_url, {})
+        self.assertEqual(response.status_code, 200)
+        self.assertJSONEqual(
+            response.content.decode(),
+            {"is_valid": False, "is_available": True},
+        )
+
+        # Check the user can still see the preview with the last valid data
+        self.assertIn(self.edit_session_key, self.client.session)
+
+        response = self.client.get(self.preview_on_edit_url)
+
+        # Check the HTML response
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, "tests/previewable_setting.html")
+        soup = self.get_soup(response.content)
```

Comment:
Now that you added span with ids, I wonder how Beautifulsoup fares in terms of performance compared to `assertContains`/`assertNotContains`/

Reply:
I'm curious too. In my experience so far, I don't see a noticeable slowdown, and the benefit of being able to have more targeted tests has definitely been a big DX boost.